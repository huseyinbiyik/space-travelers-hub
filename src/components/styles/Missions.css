.missions {
  max-width: 1200px;
  margin: 0 auto;
}

.missions table {
  margin-top: 30px;
  width: 100%;
  text-align: left;
  border-collapse: collapse;
  margin-bottom: 50px;
}

.missions thead {
  font-weight: bold;
  background-color: #ccc;
}

.missions thead th {
  padding: 10px;
}

.missions thead th:first-child {
  width: 10%;
}

.missions thead th:nth-child(2) {
  width: 66%;
}

.missions thead th:nth-child(3) {
  width: 12%;
}

.missions thead th:nth-child(4) {
  width: 12%;
}

.missions tbody tr td {
  padding: 10px;
}

.missions_name {
  font-weight: bold;
}

.missions_status {
  text-align: center;
  font-weight: 500;
  font-size: 0.7rem;
  color: white;
}

.missions_status p {
  background-color: #626060;
  padding: 4px;
  border-radius: 5px;
}

.missions_set_active p {
  background-color: dodgerblue;
}

.missions_actions {
  text-align: center;
  font-weight: 500;
}

.missions_actions button {
  cursor: pointer;
  padding: 5px;
  border-radius: 2px;
  border: 1px solid #626060;
  background-color: transparent;
  color: #626060;
}

.missions_actions_active button {
  border: 1px solid red;
  color: red;
}
