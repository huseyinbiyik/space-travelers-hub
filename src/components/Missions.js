import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions, joinMission } from '../redux/missions/missions';
import styles from './styles/Missions.css';

const Missions = () => {
  const missions = useSelector((state) => state.mission);
  const dispatch = useDispatch();
  useEffect(() => {
    if ((missions || []).length === 0) {
      dispatch(fetchMissions());
    }
  }, []);

  const toggleReservation = (mission) => {
    if (!mission.reserved) {
      dispatch(joinMission(mission.mission_id));
    }
  };

  const tableHeaders = ['Mission', 'Description', 'Status', 'Action'];

  return (
    <section className="main-wrapper">
      <table border={1} cellSpacing={0}>
        <thead>
          <tr>
            {tableHeaders.map((header) => (
              <th key={header}>{header}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {missions
            && missions.map((mission) => (
              <tr key={mission.mission_id}>
                <td className={styles.missions_name}>{mission.mission_name}</td>
                <td>{mission.description}</td>
                <td
                  className={`${styles.missions_status} ${
                    mission.reserved ? styles.missions_set_active : ''
                  }`}
                >
                  <p>{mission.reserved ? 'Active Member' : 'Not a Member'}</p>
                </td>
                <td
                  className={`${styles.missions_actions} ${
                    mission.reserved ? styles.missions_actions_active : ''
                  }`}
                >
                  <button
                    key={mission.mission_id}
                    onClick={() => toggleReservation(mission.mission_id)}
                    type="button"
                  >
                    {mission.reserved ? 'Leave Mission' : 'Join Mission'}
                  </button>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </section>
  );
};

export default Missions;
