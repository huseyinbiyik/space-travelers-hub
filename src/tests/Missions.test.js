import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import store from "../redux/configureStore";
import Missions from "./../components/Missions";
import missionsReducer, {
  showMissions,
  joinMission,
  leaveMission,
} from "../redux/missions/missions";
import testDataMissions from "./testdatamissions";

describe("Mission", () => {
  it("should render mission component", () => {
    const missions = render(
      <Provider store={store}>
        <Missions />
      </Provider>
    );
    expect(Missions).toMatchSnapshot();
  });

  it("Should update the state with API data", () => {
    expect(
      missionsReducer(testDataMissions, showMissions(testDataMissions))
    ).toEqual(testDataMissions);
  });

  it("Should update the state with correct id to reserved: true", () => {
    const mission_id = "9D1B7E0";
    expect(missionsReducer(testDataMissions, joinMission(mission_id))).toEqual(
      testDataMissions.map((d) =>
        d.id === mission_id ? { ...d, reserved: true } : d
      )
    );
  });

  it("Should update the state with correct id to reserved: false", () => {
    const mission_id = "F4F83DE";
    expect(missionsReducer(testDataMissions, leaveMission(mission_id))).toEqual(
      testDataMissions.map((d) =>
        d.id === mission_id ? { ...d, reserved: false } : d
      )
    );
  });
});
