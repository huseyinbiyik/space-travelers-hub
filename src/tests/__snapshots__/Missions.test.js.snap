// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission should render mission component 1`] = `[Function]`;
